### PR TITLE
Jzx/to host n dealloc

### PR DESCRIPTION
**Pull request by jameszianxuTT**
### Ticket
Link to Github Issue

### Problem description
Provide context for the problem.

### What's changed
Describe the approach used to solve the problem.
Summarize the changes made and its impact.

### Checklist
- [ ] New/Existing tests provide coverage for changes

### CODE REVIEW

**Comment by vmilosevic** [comment]
<table><tr><th><th>Tests</th><th>Passed ✅</th><th>Skipped ⚠️</th><th>Failed</th></tr><tr><td>TT-Torch Tests</td><td>766 ran</td><td>751 passed</td><td>15 skipped</td><td>0 failed</td></tr></table>

<table><tr><th>Test</th><th>Result</th></tr><tr><td colspan="2">No test annotations available</td></tr></table>

**Comment by jameszianxuTT** [comment]
Test executes successfully in CI: https://github.com/tenstorrent/tt-torch/actions/runs/16298377142/job/46026962396?pr=1061#step:10:41

File: tt_torch/csrc/bindings.cpp [hunk 1]
```diff
@@ -473,14 +480,15 @@ PYBIND11_MODULE(tt_mlir, m) {
         "returning the runtime tensors on device");
   m.def(
       "to_host",
-      [](py::args args) {
-        auto result = to_host(args);
+      [](py::args args, bool deallocate_tensor) {
+        auto result = to_host(args, deallocate_tensor);
```
Comment (jameszianxuTT) [inline tt_torch/csrc/bindings.cpp:484]: @hshahTT The purpose of this change is for golden verification against static cache tensors. 

Static cache tensors are allocated on device once at prefill and then their tensor handles are reused for subsequent decode steps and updated in place without being output from the graph.

One verification mode is to dump to host those on-device static cache tensors at each decode step and compare them to the static cache tensors updated by the golden model. This can allow us to track the evolution of those static cache tensors over time. But since we want to dump these runtime tensors at each step and also reuse them in the next step, I added this non-deallocating pass.

File: tests/models/llama/test_llama3_generative.py [hunk 2]
```diff
@@ -0,0 +1,342 @@
...
+    # Allow local disablement of golden verification to accelerate tests
+    # by avoiding dev2host transfer of static cache
+    enable_golden = True
+
+    # Setup compilation
+    clear_dynamo_cache()
+    cc = CompilerConfig()
+
+    # Consteval disabled due to 4D Causal Attention Mask evaluation getting constant folded in torchfx
+    #   due to incorrect tracing of static cache and malformed output missing static cache tensors
```
Comment (hshahTT) [inline tests/models/llama/test_llama3_generative.py:90]: Should we open an issue to track this?

File: tt_torch/csrc/bindings.cpp (hunk 1 above)
Comment (hshahTT) [reply tt_torch/csrc/bindings.cpp:484]: Sounds good

**Review by hshahTT — APPROVED** [review]

**Comment by jameszianxuTT** [comment]
Local debug utils look like this:
image

File: tests/models/llama/test_llama3_generative.py [hunk 3]
```diff
@@ -0,0 +1,342 @@
...
+        print(f"  Max time: {max(decode_times):.3f}s")
+        print(f"  Average tokens/second: {len(decode_times) / sum(decode_times):.2f}")
+        print()
+
+    if enable_golden and golden_pccs and any(pcc > 0 for pcc in golden_pccs):
+        print(f"ACCURACY:")
+        average_pcc = sum(golden_pccs) / len(golden_pccs)
+
+        assert (
+            average_pcc >= 0.6
```
Comment (jameszianxuTT) [inline tests/models/llama/test_llama3_generative.py:288]: This is kind of arbitrary, it checks that across all tokens generated, the logits-to-logits pcc averages to > 0.6 

In practise, I see a lot of prompt-dependent variation where the output still makes "sense". This has an "average PCC" of 0.768

```
Initial prompt: '<|begin_of_text|>I like taking walks in the'
Generated text: '<|begin_of_text|>I like taking walks in the woods. I like the way the trees look, the way the sun shines through them, the way the wind blows through them. I like the way the birds'
Golden text: '<|begin_of_text|>I like taking walks in the woods. I like the smell of the trees, the sound of the birds, the feel of the air. I like the way the sun shines through the leaves'
```

**Review by AleksKnezevic — APPROVED** [review]
